re2: change to cmake build  to support gprc, which currently needs a cmake version of re2 to build. Try to make it work like the previous autotools build worked.

essandess opened the pull request:
re2: Add cmake build and .cmake files
* Fixes: https://trac.macports.org/ticket/62168#ticket
* Related: https://github.com/macports/macports-ports/pull/7791

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
printf "%s\n" "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)" "$(xcodebuild -version|awk 'NR==1{x=$0}END{print x" "$NF}')"|tee /dev/tty|pbcopy
-->
macOS 10.15.7 19H114
Xcode 12.4 12D4e

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

kencu commented:
You need to turn on the shared libs, which is off by default:
```
option(BUILD_SHARED_LIBS "build shared libraries" OFF)
```
otherwise we get the static libs, which is not what we currently have, and not what we want:
```
2021-01-28T06:12:49.7004010Z x ./opt/
2021-01-28T06:12:49.7008140Z x ./opt/local/
2021-01-28T06:12:49.7010060Z x ./opt/local/include/
2021-01-28T06:12:49.7011530Z x ./opt/local/lib/
2021-01-28T06:12:49.7013120Z x ./opt/local/share/
2021-01-28T06:12:49.7014780Z x ./opt/local/share/doc/
2021-01-28T06:12:49.7016700Z x ./opt/local/share/doc/re2/
2021-01-28T06:12:49.7027570Z x ./opt/local/share/doc/re2/syntax.html
2021-01-28T06:12:49.7033670Z x ./opt/local/share/doc/re2/LICENSE
2021-01-28T06:12:49.7038700Z x ./opt/local/share/doc/re2/AUTHORS
2021-01-28T06:12:49.7047190Z x ./opt/local/share/doc/re2/syntax.txt
2021-01-28T06:12:49.7053940Z x ./opt/local/share/doc/re2/README
2021-01-28T06:12:49.7058860Z x ./opt/local/share/doc/re2/CONTRIBUTORS
2021-01-28T06:12:49.7060500Z x ./opt/local/lib/cmake/
2021-01-28T06:12:49.7281760Z x ./opt/local/lib/libre2.a
2021-01-28T06:12:49.7283480Z x ./opt/local/lib/cmake/re2/
2021-01-28T06:12:49.7290110Z x ./opt/local/lib/cmake/re2/re2Config-macports.cmake
2021-01-28T06:12:49.7295220Z x ./opt/local/lib/cmake/re2/re2Config.cmake
2021-01-28T06:12:49.7297730Z x ./opt/local/include/re2/
2021-01-28T06:12:49.7304000Z x ./opt/local/include/re2/stringpiece.h
2021-01-28T06:12:49.7309760Z x ./opt/local/include/re2/filtered_re2.h
2021-01-28T06:12:49.7324920Z x ./opt/local/include/re2/re2.h
2021-01-28T06:12:49.7331010Z x ./opt/local/include/re2/set.h
```

Steve -- this is stuff you should know by now.

kencu reviewed (changes requested):
use shared libs

kencu commented:
```
2021-02-01T12:24:25.6142820Z x ./+STATE
2021-02-01T12:24:25.6148240Z x ./+PORTFILE
2021-02-01T12:24:25.6154620Z x ./+CONTENTS
2021-02-01T12:24:25.6158230Z x ./+DESC
2021-02-01T12:24:25.6160730Z x ./+COMMENT
2021-02-01T12:24:25.6162480Z x ./opt/
2021-02-01T12:24:25.6167060Z x ./opt/local/
2021-02-01T12:24:25.6169820Z x ./opt/local/include/
2021-02-01T12:24:25.6172090Z x ./opt/local/lib/
2021-02-01T12:24:25.6174250Z x ./opt/local/share/
2021-02-01T12:24:25.6176850Z x ./opt/local/share/doc/
2021-02-01T12:24:25.6179980Z x ./opt/local/share/doc/re2/
2021-02-01T12:24:25.6194180Z x ./opt/local/share/doc/re2/syntax.html
2021-02-01T12:24:25.6200720Z x ./opt/local/share/doc/re2/LICENSE
2021-02-01T12:24:25.6207200Z x ./opt/local/share/doc/re2/AUTHORS
2021-02-01T12:24:25.6215540Z x ./opt/local/share/doc/re2/syntax.txt
2021-02-01T12:24:25.6221120Z x ./opt/local/share/doc/re2/README
2021-02-01T12:24:25.6226530Z x ./opt/local/share/doc/re2/CONTRIBUTORS
2021-02-01T12:24:25.6374480Z x ./opt/local/lib/libre2.dylib
2021-02-01T12:24:25.6376240Z x ./opt/local/lib/cmake/
2021-02-01T12:24:25.6379090Z x ./opt/local/lib/cmake/re2/
2021-02-01T12:24:25.6386650Z x ./opt/local/lib/cmake/re2/re2Config-macports.cmake
2021-02-01T12:24:25.6393370Z x ./opt/local/lib/cmake/re2/re2Config.cmake
2021-02-01T12:24:25.6395410Z x ./opt/local/include/re2/
2021-02-01T12:24:25.6402830Z x ./opt/local/include/re2/stringpiece.h
2021-02-01T12:24:25.6410450Z x ./opt/local/include/re2/filtered_re2.h
2021-02-01T12:24:25.6426850Z x ./opt/local/include/re2/re2.h
2021-02-01T12:24:25.6432100Z x ./opt/local/include/re2/set.h
```

I see now we have this:
```
/opt/local/lib/libre2.dylib
```

but now that is completely different than what the port previously provided.
```
  /opt/local/lib/libre2.9.0.0.dylib
  /opt/local/lib/libre2.9.dylib
  /opt/local/lib/libre2.a
  /opt/local/lib/libre2.dylib
```

So now that all has to get straightened out.

We'll need the versioned libs, and (it appears) the archive lib as well.

Otherwise, everything that uses re2 will have to be individually tested and built to see if it still builds right, and then, if it all does, everything will need to be rev-bumped, probably...

kencu commented:
no to mention actually inspecting the libs to see if the compatibility version, etc, has changed.

kencu commented:
ArchLinux went through this process a few months ago and found other problems as well.

Perhaps you might get inspiration from here <https://bugs.archlinux.org/task/67739>

kencu commented:
eg the pkg-config file is gone now, from the cmake build, and we'll need that back.

essandess commented:
Thanks for the (obscure!) pointers and fixes. Latest push addresses all your points.

kencu commented:
NB:

"RE2 supports CMake mostly for the sake of users who don't have GNU make (e.g. Visual Studio users, Xcode users) and only partly – and unenthusiastically – for the sake of CMake users."

"Perhaps it's better not to add anything to re2 and rely on downstream projects to use the pkg-config module for CMake."

When this identical issue came up in ArchLinux it appears the final outcome was that gprc learned how to properly use pkgconfig modules, and re2 was left alone:

<https://github.com/google/re2/issues/274>.


So -- In the end, cmake is not properly supported by re2, they don't ever plan to properly support it, and I don't see this being the right approach for re2 here. I would suggest you go through the efforts they did of making gprc use pkg-config files.

kencu commented:
@essandess  -- just to understand the logic here.

There are a number of packages in MacPorts, including the qt5 series, that use re2, and of course who knows how many people use the macports build of it for other projects.

It has to be reliable for all of them, not just suit an update to grpc.

essandess commented:
Thanks @kencu.

I confirm a successful build of the latest version gf `grpc` using the existing MacPorts install.

Therefore, I will close this PR.

essandess commented:
@kencu A `cmake` build of `re2` Is required to build the `py*-pyarrow` ports in #7791 .

I’m reopening this PR to provide the cmake build, which is critical path for several other PR’s that depend upon `pyarrow`. 

See:
* https://issues.apache.org/jira/browse/ARROW-11888?focusedCommentId=17300689&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17300689
* https://github.com/grpc/grpc/issues/25434

kencu commented:
It would be helpful if someone demonstrated that the major existing consumers of re2, esp qt5*-qtwebengine, build and work properly against this version.

essandess commented:
> It would be helpful if someone demonstrated that the major existing consumers of re2, esp qt5*-qtwebengine, build and work properly against this version.

This issue blocks checking these `qt5` builds: https://trac.macports.org/ticket/62433

The `re2` libraries and versioning is exactly the same between the Makefile and cmake builds, and the cmake build works with other builds. Is there any reason to believe that this cmake build wouldn't work with `qt5*-qtwebengine` or others?

kencu commented:
dunno. last time I looked the build was completely different than the autotools build, the libraries had all different names, no idea about the library version numbers, and I didn't know what was happening with the pkgconfig files at all.

You fixed all that?

kencu commented:
My basic issue here is that 100% of MacPorts users (pretty much) install qt5 and require it to work, and an important part of it appears to rely on re2.

And I know you'd like to build Apache Arrow in MacPorts, and perhaps it's useful to some people, but I'm going to guess it will be like 1-2% of users.

So whatever we do to fix things up for Apache Arrow, we cannot break qt5 in the process.

This is only my opinion. Others may feel it's totally fine, and we'll fix the wreckage, if any.

essandess commented:
> dunno. last time I looked the build was completely different than the autotools build, the libraries had all different names, no idea about the library version numbers, and I didn't know what was happening with the pkgconfig files at all.
> 
> You fixed all that?

I believe so:
```bash
port contents  re2
Port re2 contains:
  /opt/local/include/re2/filtered_re2.h
  /opt/local/include/re2/re2.h
  /opt/local/include/re2/set.h
  /opt/local/include/re2/stringpiece.h
  /opt/local/lib/cmake/re2/re2Config-macports.cmake
  /opt/local/lib/cmake/re2/re2Config.cmake
  /opt/local/lib/libre2.9.0.0.dylib
  /opt/local/lib/libre2.9.dylib
  /opt/local/lib/libre2.dylib
  /opt/local/lib/pkgconfig/re2.pc
  /opt/local/share/doc/re2/AUTHORS
  /opt/local/share/doc/re2/CONTRIBUTORS
  /opt/local/share/doc/re2/LICENSE
  /opt/local/share/doc/re2/README
  /opt/local/share/doc/re2/syntax.html
  /opt/local/share/doc/re2/syntax.txt
```

kencu commented:
Thanks, Steve -- 

I guess really, in the end of it all, these two files are all you really want, right?
```
  /opt/local/lib/cmake/re2/re2Config-macports.cmake
  /opt/local/lib/cmake/re2/re2Config.cmake
```

lets see how the pkgconfig files compare, and what the dylib version numbers look like, etc. If they all match with what is there now, that's a step.

I guess we'd have to sacrifice the static lib as it's very hard to make both a dylib and a static lib with a cmake build, although some builds seem to have figured it out. No idea if anything wants that, but it's not MacPorts standard to give it anyway I suppose.

kencu commented:
the existing lib is this:
```
% otool -L /opt/local/lib/libre2.9.0.0.dylib
/opt/local/lib/libre2.9.0.0.dylib:
	/opt/local/lib/libre2.9.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```
the new cmake lib is this:
```
 % otool -L libre2.9.0.0.dylib
libre2.9.0.0.dylib:
	/opt/local/lib/libre2.9.dylib (compatibility version 9.0.0, current version 9.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
```
both of those versions are wrong. Arguably the 9.0.0 version is really wrong, but who's to quibble.


the existing pkgconfig file is this:
```
% cat /opt/local/lib/pkgconfig/re2.pc
includedir=/opt/local/include
libdir=/opt/local/lib

Name: re2
Description: RE2 is a fast, safe, thread-friendly regular expression engine.
Version: 0.0.0
Cflags: -std=c++11 -pthread -I${includedir}
Libs: -pthread -L${libdir} -lre2
```

the cmake build one is this:
```
% cat re2.pc
includedir=/opt/local/include/re2
libdir=/opt/local/lib

Name: re2
Description: RE2 is a fast, safe, thread-friendly regular expression engine.
Version: 0.0.0
Cflags: -std=c++11 -pthread -I${includedir}
Libs: -pthread -L${libdir} -lre2
```

the cmake version is wrong in both, but for some reason is 0.0.0 in the pkgconfig, whereas the library is version 9.0.0. 

The header locations are different. 

I have no idea if any of this will matter to the software that currently uses re2. I have not built qt5*-qtwebengine against these new versions, although I did spend several weeks fixing and building qt5*-qtwebengine in December to work on all the newer systems.

Really, I think this is all for you to fix up and spend the time testing to be sure it's all OK rather than me Steve.

essandess commented:
> My basic issue here is that 100% of MacPorts users (pretty much) install qt5 and require it to work, and an important part of it appears to rely on re2.

@kencu

I confirm that `sudo port -vst install qt5-qtwebengine` builds, installs, and works with this PR.

[What a monster! This required setting `sudo launchctl limit maxfiles 256000 2097152` to get `ninja` to compile this thing.]

I do not foresee any issues merging this PR.

essandess commented:
> And I know you'd like to build Apache Arrow in MacPorts, and perhaps it's useful to some people, but I'm going to guess it will be like 1-2% of users.

I wouldn’t count on that. Apache Arrow is in the critical path of some widely used deep learning frameworks. And ML on the Mac just became a whole lot [more interesting](https://blog.tensorflow.org/2020/11/accelerating-tensorflow-performance-on-mac.html). That’s why we want Apache Arrow, in addition to its basic large data capabilities.

cjones051073 commented:
So, ready the comments it seems we might be ready to go on this, as all the issues are (mostly) addressed ?

essandess commented on devel/re2/Portfile:
```diff
@@ -22,23 +23,34 @@ checksums           rmd160  47471514269abd3b815358fa2e8934d4489985e3 \
+# https://bugs.archlinux.org/task/67739
+# https://gist.github.com/mtorromeo/1d48de16534dc8ee29cd872f94b070e5
+
+pre-configure {
+    reinplace s|@includedir@|${prefix}/include/${name}|g \
+        ${worksrcpath}/re2.pc
+    reinplace s|@libdir@|${prefix}/lib|g \
+        ${worksrcpath}/re2.pc
+    set somajor [exec /bin/sh -c "grep '^SONAME=' ${worksrcpath}/Makefile | cut -d= -f2"]
+    # append this line to CMakeLists.txt
+    reinplace "\$ a\\
+set_target_properties(re2 PROPERTIES VERSION ${somajor}.0.0 SOVERSION ${somajor})" \
+        ${worksrcpath}/CMakeLists.txt
+}
```
Comment: this adds shared library versioning per:
* https://bugs.archlinux.org/task/67739
* https://gist.github.com/mtorromeo/1d48de16534dc8ee29cd872f94b070e5

cjones051073 commented:
@kencu are you happy now to go on this ?

kencu commented:
haven't built it. I had some concerns, as above with the different build products etc, but if it is verified that qt5 is fine with this, that is a bonus. I put a lot of time into getting qt5 working in Dec/Jan. 

And to be honest, qt5 is @MarcusCalhoun-Lopez 's port, and he's back again now, so if there is damage he will be called in to fix it not me.

For a lark, I built Apache Arrow without this PR -- it uses "Findre2Alt.cmake" to use the pkg-config files, like they do on linux builds, and found and linked against re2 without any trouble.

The python modules -- well I can't tell if it did or didn't find re2 to be honest. Maybe they didn't, like Steve said. I don't know what the linux builds do about the python modules and re2.

kencu commented:
Not many (any) of the linux distros are doing this. They are all telling cmake-building sofware projects to use this instead:

```
find_package(PkgConfig)
pkg_check_modules(RE2 re2)
if(RE2_FOUND)
  add_library(re2::re2 INTERFACE IMPORTED)
  set_target_properties(re2::re2 PROPERTIES
    INTERFACE_COMPILE_OPTIONS "${RE2_CFLAGS}"
    INTERFACE_LINK_LIBRARIES "${RE2_LDFLAGS}"
    )
endif()
```

But -- all I really cared about was qt5, and that is not really my issue any more.

essandess commented:
> For a lark, I built Apache Arrow without this PR -- it uses "Findre2Alt.cmake" to use the pkg-config files, like they do on linux builds, and found and linked against re2 without any trouble.

Did you build it with the latest version of grpc 1.36.2 in #9729

We must migrate grpc to a cmake build, and I found that building Apache Arrow with updated grpc was a major PITA and ran into issues that a lot of others are reporting.

essandess commented:
@kencu @cjones051073 

> Not many (any?) of the linux distros are doing this, mostly because (it appears) cmake has great, unusual, weirdly troublesome difficulty building both a dylib and a static lib without great shenanigans.
> 
> They are all telling cmake-using sofware projects to use something like this instead:

This is exactly how Homebrew builds apache-arrow, grpc, and re2, all cmake builds:
* https://formulae.brew.sh/formula/apache-arrow
* https://formulae.brew.sh/formula/grpc
* https://formulae.brew.sh/formula/re2

Because this is what works. Homebrew has been providing these ports and their recent updates for much longer than we have. #7791 was submitted nearly 9 months ago… Homebrew has providing apache-arrow since [2017](https://github.com/Homebrew/homebrew-core/commits/dbe01165ab730071d0da276649562308b3928e0d?after=dbe01165ab730071d0da276649562308b3928e0d+69&branch=dbe01165ab730071d0da276649562308b3928e0d&path%5B%5D=Formula&path%5B%5D=apache-arrow.rb). There's a lesson here. We shouldn't work harder to review a submission than it takes to address any issues it has and spiral forward.

My recommendation is to do what works as soon as possible, and move on to the higher-level ports that use these PRs.

kencu commented:
anyone can commit. As I said before, it is not up to me what gets committed. I only review what interests my and spot issues (that helped you fix this up considerably, I might add).

You have to defend your commits to @ryandesign and I'm not prepared to do that for this commit, but @cjones051073 can do whatever he likes.

kencu commented:
besides, homebrew doesn't do quite this. they do this:

```
  def install
    ENV.cxx11

    # Run this for pkg-config files
    system "make", "common-install", "prefix=#{prefix}"

    # Run this for the rest of the install
    system "cmake", ".", "-DBUILD_SHARED_LIBS=ON", "-DRE2_BUILD_TESTING=OFF", *std_cmake_args
    system "make"
    system "make", "install"
  end
```

Perhaps "make common-install" gives them all the stuff we need. Who knows?

I'm signing off here.

essandess commented:
@kencu 

> I only review what interests me and spot issues (that helped you fix this up considerably, I might add).

Indeed, your help was both invaluable and impressive in tracking down the obscure links required to fix this—thank you very much.

> Perhaps "make common-install" gives them all the stuff we need. Who knows?

The `common-install` target only installs the pkgconfig file with our addressing all the issues you identify and correct above. See:

https://github.com/google/re2/blob/f8e389f3acdc2517562924239e2a188037393683/Makefile#L316-L321

To the best of my knowledge, this PR works, addresses all of @kencu's points (except for the upstream versioning issues; see https://github.com/google/re2/issues/304), will build qt5, and provides a path to building the downstream ML frameworks that use `re2`.

cjones051073 commented:
I am going to merge this, just to unblock things in other PRs.

essandess commented:
@kencu You get the credit when all these issues get fixed in the forthcoming upstream updates discussed in https://github.com/google/re2/issues/304.
